update_annotations_from_table.py

```diff
@@ -0,0 +1,207 @@
+ch = logging.StreamHandler()
+ch.setLevel(logging.DEBUG)
+ch.setFormatter(formatter)
+logger.addHandler(ch)
+
+from multiprocessing.dummy import Pool
+
+import synapseclient
+
+syn = synapseclient.Synapse(skip_checks=True)
+syn.login(silent=True)
+
+# Taken from https://github.com/larssono/Analysis_helpers/blob/master/synapseHelpers.py
+# Should probably be in synapseclient.synapseutils
+SYNAPSE_PROPERTIES = ['benefactorId', 'nodeType', 'concreteType', 'createdByPrincipalId',
+                      'createdOn', 'createdByPrincipalId', 'eTag', 'id', 'modifiedOn',
+                      'modifiedByPrincipalId', 'noteType', 'versionLabel', 'versionComment',
+                      'versionNumber', 'parentId', 'description', 's3Token', 'name']
+
+def query2df(queryContent, filterSynapseFields=True, savedSynapseFields = ('id', 'name')):
+    """Converts the returned query object from Synapse into a Pandas DataFrame.
+
+    Arguments:
+    - `queryContent`: content returned from query or chunkedQuery
+    - `filterSynapseFields`: Removes Synapse properties of the entity returned with "select * ..."
+                             defaults to True
+    - `savedSynapseFields`: list of synapse fields not to filter defaults to id, name
+
+    """
+
+    removedProperties = set(SYNAPSE_PROPERTIES) - set(savedSynapseFields)
+
+    try:
+        queryContent = queryContent['results']
+    except TypeError: #If 'results' not present it is a geneartor
+        queryContent = list(queryContent)
+    for row in queryContent:
+        for key in row.keys():
+            prefix, new_key = key.split('.', 1)
+            item = row.pop(key)
+            row[new_key] = item[0] if type(item) is list else item
+            if filterSynapseFields and new_key in removedProperties:
+                del row[new_key]
+    return pd.DataFrame(queryContent)
+
+def fixDict(d):
+    """Fix dictionaries having lists and nans.
+
+    """
+
+    newd = {}
+
+    for k, v in d.iteritems():
+        tmp = {}
+
+        for k2, v2 in v.iteritems():
+
+            if v2 is list:
+                tmp[k2] = v2
+            else:
+                try:
+                    if not numpy.isnan(v2):
+                        tmp[k2] = v2
+                except TypeError:
+                    tmp[k2] = v2
+
+        newd[k] = tmp
+
+    return newd
+
+def updateDict(a, b, removeFields=SYNAPSE_PROPERTIES):
+    """Remove synapse fields from Synapse query on file annotations and update with a second dictionary.
+
+    """
+
+    foo = [a.pop(x) for x in removeFields]
+    a.update(b)
+    return a
+
+def updateAnnots(synid, lookupDict, syn):
+    """Update annotations using a dict to lookup which annotations to add (keys are synids).
+
+    Works this way for parallelization.
+
+    """
+
+    logger.info("Updating annotations for %s" % synid)
+
+    o = syn.get(synid, downloadFile=False)
+    a = syn.getAnnotations(o)
+
+    # Make a copy so updates do not affect the original
+    newa = a.copy()
```

Comment:
why making a copy? seems that I shouldn't, but I think it was due to something in `updateDict`.